### PR TITLE
Cluster: Use per-member server certs

### PR DESCRIPTION
**Pull request by tomponline**
- Introduces concept of `servertCert` and `networkCert` to represent the certificate used for each server's client certificate when making intra-cluster requests, and the certificate used when exposing the API onto the network respectively.
- Notifies other cluster members when a trusted certificate is updated or deleted, to allow their local trusted certificate caches to be refreshed.
- Requires and ensures that trusted server certificates are stored with a name of `<server name>` and a type of `server`. This is the server certificate for a cluster member can be found, and that they are trusted correctly as server certificates.
- Adds a `certificates` table to the local DB which caches the server type certificates from the global DB, to allow the LXD daemon to re-establish a connection to the cluster when starting up.
- Detects if the local `certificates` table is empty, and if so switches to pre-server-cert mode, that uses the network certificate for both the client certificate and the trusted certificate for incoming requests. This allows the cluster communication to be established allowing the per-server certificate patch to run,
- Adds a patch that will wait add the server certificate to the global trust store, and wait until all other cluster members have done the same. Once this is completed, the LXD daemon will load the trusted certificates into its cache and switch to using its own server certificate for intra-cluster communication.
- When a member is removed from the cluster, any trusted server certificates matching the member's name are removed from the database and the cached trusted certificates are refreshed. This ensures that once a member is removed it can no longer access the cluster.
- Updates the heartbeat handler to refresh cached trusted certificates from the global DB when the member list changes (added or removed) to ensure that we only trust active members in the cluster.

### CODE REVIEW

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@stgraber this is ready for review, however the jenkins tests are failing since this morning when I added additional safety checks to ensure that when joining a cluster that there isn't an existing trusted server certificate that matches the joining member's server certificate, but recorded under a different name (to avoid forcefully replacing an existing server certificate).

This situation doesn't occur normally, but in the test suite it only currently has two server certificates available, whereas the tests themselves spawn up to 8 different members. So I need to add 7 more certificate key pairs and update the spawn function to use a different certificate for each member.

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
@tomponline i386 is consistently broken. I suspect that's because of them running on Ubuntu 18.04.
The issue seems to be around how certs were generated and lack of SAN fields.

**Comment by tomponline** [comment]
@stgraber Yes indeed, I've been looking into this today and have found the following items:

1. The easy-rsa version 2 in Bionic doesn't support custom Subject Alternative Names.
2. Even on Focal 64bit with a more recent version of easy-rsa present, if I run the PKI test in isolation, I get the same error.

This leads me to think that the LXD_ALT_CERT=1 env var was somehow leaking between tests, effectively causing the `spawn_lxd` function to replace the pkitool/easy-rsa generated server cert with the pre-generated alternative server cert, which I've confirmed contains an IP SAN. And when it was run in isolation with LXD_ALT_CERT unset, the `spawn_lxd` function was replacing the pkitool/easy-rsa generated server cert with the pre-generated main server cert, which I've confirmed does not contain an IP SAN.

As you can see here:
https://github.com/lxc/lxd/blob/master/test/includes/lxd.sh#L29-L36

This is no scenario where pre-existing server.crts are not replaced by one of the pre-generated ones when calling `spawn_lxd` as the PKI tests do:

https://github.com/lxc/lxd/blob/master/test/suites/pki.sh#L43-L47

So based on your comment https://github.com/lxc/lxd/pull/8676#discussion_r614738225 and the fact the tests have been historically using a predefined cert and not one signed by the CA, I am going to do the following:

1. Remove the LXD_ALT_CERT concept entirely, and always get LXD to generate its own certificates. As this will better test LXD in a real-world scenario and prevent leaking cert state between tests.
2. Remove the pre-generated alternative certificate. They will not be used any more.
3. Leave the pre-generated main certificate. It will not be used by the test suite any more, but is still used by the go unit tests.
4. Don't copy pkitool/easy-rsa server cert, as its not used now, and we don't want/need it to be used once the LXD_ALT_CERT concept is removed.

**Comment by tomponline** [comment]
@stgraber also I am investigating the CA certificate revocation check functionality, as it seems to the current tests are passing but they don't properly test that revoked certificates without a password are blocked (which is my understanding of what should happen).

**Comment by tomponline** [comment]
@stgraber OK so the tests are now passing for i386, but https://github.com/lxc/lxd/pull/8719 has confirmed my suspicion that the CRL checks are not working as intended on main (even without this PR).

**Comment by tomponline** [comment]
@stgraber this test run includes https://github.com/lxc/lxd/pull/8719

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/3435/ looking pretty green

**Comment by tomponline** [comment]
@stgraber I've seen ceph go green before, so should be ok :)

File: lxd/certificates.go [hunk 1]
```diff
@@ -164,17 +164,20 @@ func certificatesGet(d *Daemon, r *http.Request) response.Response {
 }
 
 func updateCertificateCache(d *Daemon) {
+	logger.Debug("Refreshing trusted certificate cache")
```
Comment (stgraber) [inline lxd/certificates.go:168]: Would have been nice to split the logging/cleanup bits into a separate commit for easier cherry-picking.
But I'm currently planning on cherry-picking all this work anyway, so not actually needed.

File: lxd/certificates.go [hunk 2]
```diff
@@ -226,6 +226,52 @@ func updateCertificateCache(d *Daemon) {
 	d.clientCerts.Lock.Unlock()
 }
 
+// updateCertificateCacheFromLocal loads trusted server certificates from local database into memory.
+// If no local trusted server certificatds available and is clustered, then loads network certificate into trusted
```
Comment (stgraber) [inline lxd/certificates.go:231]: type here s/certificatds/certificates/

Comment (tomponline) [reply lxd/certificates.go]: fixed

File: lxd/daemon.go [hunk 3]
```diff
@@ -875,20 +875,54 @@ func (d *Daemon) init() error {
...
+	if err != nil {
+		return err
+	}
+
+	clustered, err := cluster.Enabled(d.db)
+	if err != nil {
+		return errors.Wrapf(err, "Failed checking if clustered")
+	}
+
+	// Detect if clusterd, but not yet upgraded to per-server client certificates.
```
Comment (stgraber) [inline lxd/daemon.go:897]: *clustered

Comment (tomponline) [reply lxd/daemon.go]: fixed

**Pull request merged**